src/time_stream/infill.py

```diff
@@ -0,0 +1,356 @@
+        """Factory method to get an infill method class instance from string names or class type.
+
+        Args:
+            method: The infill method specification, which can be:
+                - A string name: e.g. "linear_interpolation"
+                - A class type: LinearInterpolation
+                - An instance: LinearInterpolation(), or any InfillMehod instance
+            **kwargs: Parameters specific to the infill method, used to initialise the class object.
+                      Ignored if method is already an instance of the class.
+
+        Returns:
+            An instance of the appropriate InfillMethod subclass.
+
+        Raises:
+            KeyError: If a string name is not registered as an infill method.
+            TypeError: If the input type is not supported or a class doesn't inherit from InfillMethod.
+        """
+        # If it's already an instance, return it
+        if isinstance(method, InfillMethod):
+            return method
+
+        # If it's a string, look it up in the registry
+        if isinstance(method, str):
+            try:
+                return _INFILL_REGISTRY[method](**kwargs)
+            except KeyError:
+                raise KeyError(f"Unknown infill method '{method}'.")
+
+        # If it's a class, check the subclass type and return
+        elif isinstance(method, type):
+            if issubclass(method, InfillMethod):
+                return method(**kwargs)  # type: ignore[misc]
+            else:
+                raise TypeError(f"Infill method class {method.__name__} must inherit from InfillMethod")
+
+        else:
+            raise TypeError(f"Infill method must be a string or an InfillMethod class. Got {type(method).__name__}")
+
+    def apply(
+        self,
+        ts: TimeSeries,
+        infill_column: str,
+        observation_interval: Optional[datetime | Tuple[datetime, datetime | None]] = None,
+        max_gap_size: Optional[int] = None,
+    ) -> "TimeSeries":
+        """Apply the infill method to the TimeSeries.
+
+        Args:
+            ts: The TimeSeries to check.
+            infill_column: The column to infill data within.
+            observation_interval: Optional time interval to limit the infilling to.
+            max_gap_size: The maximum size of consecutive null gaps that should be filled. Any gap larger than this
+                          will not be interpolated and will remain as null.
+        Returns:
+            TimeSeries: The infilled time series
+        """
+        # Validate column exists
+        if infill_column not in ts.columns:
+            raise KeyError(f"Infill column '{infill_column}' not found in TimeSeries.")
+
+        # Set the timeseries property
+        self._ts = ts
+
+        # We need to make sure the data is padded so that missing time steps are filled with nulls
+        df = pad_time(ts.df, ts.time_name, ts.periodicity)
```

Comment:
If the TS object has been created with `pad=true` then I guess we dont need to do this?